src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/OffenderKeyDatesService.kt

```diff
@@ -5,18 +5,40 @@ import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Transactional
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.enumerations.ReleaseDateType
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.exceptions.CrdWebException
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.CalculationSource
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.NomisCalculationSummary
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.OffenderKeyDates
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.OffenderReleaseDates
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.ReleaseDate
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.repository.CalculationRequestRepository
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.resource.CalculationController
 
 @Service
 @Transactional(readOnly = true)
 class OffenderKeyDatesService(
   private val prisonService: PrisonService,
   private val calculationResultEnrichmentService: CalculationResultEnrichmentService,
+  private val calculationRequestRepository: CalculationRequestRepository,
 ) {
 
+  fun getKeyDatesByCalcId(calculationRequestId: Long): OffenderReleaseDates {
+    try {
+      val calculationRequest = calculationRequestRepository.findById(calculationRequestId).get()
+      val offenderKeyDatesEither = prisonService.getOffenderKeyDates(calculationRequest.bookingId)
+      val releaseDatesForBookingId = offenderKeyDatesEither.map { releaseDates(it) }.getOrNull()
+      return OffenderReleaseDates(
+        calculationRequest.bookingId,
+        calculationRequestId,
+        calculationRequest.calculatedAt,
+        calculationRequest.reasonForCalculation?.displayName ?: "Not entered",
+        CalculationSource.CRDS,
+        calculationResultEnrichmentService.addDetailToCalculationDates(releaseDatesForBookingId!!, null, null).values.toList(),
```

Comment:
this is an unsafe null check, would be better to handle the error directly and throw a relevant exception. What does detailed calculation results do in this case? 
You could also make use of the either like
`offenderKeyDatesEither
       .map { releaseDates(it) }
       .map { calculationResultEnrichmentService.addDetailToCalculationDates(releaseDatesForBookingId!!, null, null).values.toList() }
      .fold(
{error -> throw SomeException(error)},
{enrichedDates -> OffenderReleaseDates(...snip...)}
)`

Reply:
Changed as Requested